### ISSUE TITLE
When querying with ExclusiveStartKey the page is identical to the previous one

### ISSUE THREAD

**Issue by ffjdm** [getmoto/moto#7888]
**Working version:** 4.2.14
**Failing version:** 5.0.11

**Scenario**
1. create an indexed table with 5 items
2. query the table with a limit equal to or greater  than the number of items -> it works and returns 1 page
3. query the table with a limit equal to the number of items - 1 -> it works and returns 2 pages
4. query the table with a limit lesser than the number of items - 1 -> it fails and returns 2 identical pages
This means if it were looping then that would never end because the same page would be fetched and the LastEvaluatedKey would remain the same.

**Expected behavior**
When querying a second time using LastEvaluatedKey as value for ExclusiveStartKey, the next items should be returned.

**Reproduction test case with 5 items**
You can modify the LIMIT constant to see the difference in behavior.
Thank you for your help 🙏 

```python
import logging
import uuid
from datetime import datetime, timezone

import boto3
from moto import mock_aws

REGION = "us-west-2"
TABLE_NAME = "test_table"
KEY_NAME = "id"
HASH_NAME = "myCriteria"
HASH_VALUE = "abc"
RANGE_NAME = "createdAt"
INDEX_NAME = f"FC_{HASH_NAME.upper()}_{RANGE_NAME.upper()}_IDX"
LIMIT = 3


@mock_aws
def test_last_key_issue():
    # Setting up the DB
    client = boto3.client("dynamodb", REGION)
    client.create_table(
        AttributeDefinitions=[
            {"AttributeName": KEY_NAME, "AttributeType": "S"},
            {"AttributeName": HASH_NAME, "AttributeType": "S"},
            {"AttributeName": RANGE_NAME, "AttributeType": "S"},
        ],
        TableName=TABLE_NAME,
        KeySchema=[
            {"AttributeName": KEY_NAME, "KeyType": "HASH"},
        ],
        GlobalSecondaryIndexes=[
            {
                "IndexName": INDEX_NAME,
                "KeySchema": [
                    {"AttributeName": HASH_NAME, "KeyType": "HASH"},
                    {"AttributeName": RANGE_NAME, "KeyType": "RANGE"},
                ],
                "Projection": {
                    "ProjectionType": "INCLUDE",
                    "NonKeyAttributes": [KEY_NAME],
                },
            },
        ],
        BillingMode="PAY_PER_REQUEST",
    )

    for idx in range(0, 5):
        client.put_item(
            TableName=TABLE_NAME,
            Item={
                "id": {"S": uuid.uuid4().hex},
                HASH_NAME: {"S": HASH_VALUE},
                "createdAt": {"S": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")},
                "customer": {"S": "mycust"},
                "idx": {"N": str(idx)},
            },
        )

    # Retrieving data
    query_params = {
        "TableName": TABLE_NAME,
        "IndexName": INDEX_NAME,
        "Select": "SPECIFIC_ATTRIBUTES",
        "ProjectionExpression": "id",
        "Limit": LIMIT,
        "ConsistentRead": False,
        "ExpressionAttributeValues": {
            ":hash": {
                "S": HASH_VALUE,
            },
        },
        "KeyConditionExpression": f"{HASH_NAME} = :hash",
        "ScanIndexForward": False,
    }

    response = client.query(**query_params)
    logging.info(response)

    start_key = response.get("LastEvaluatedKey")
    if start_key:
        query_params["ExclusiveStartKey"] = start_key
        response = client.query(**query_params)
        logging.info(response)

```

**Comment by DrGFreeman** [comment]
I've been experiencing a similar issue since version 5.0.10 was released, unfortunately I didn't have time to report it until now.

One point I may add to the above is that the result of the query with `ExclusiveStartKey` (what items it includes) depends on the sorting of the primary key values of the items as the tests below demonstrate.

<details>
<summary><b>Test code</b></summary>

```python
from uuid import uuid4

import boto3
from boto3.dynamodb.conditions import Key
from moto import mock_aws
import pytest

sorted_pk_data = [
    dict(pk="pk_A", gsi_pk="gsi_A", gsi_sk="1"),
    dict(pk="pk_B", gsi_pk="gsi_A", gsi_sk="2"),
    dict(pk="pk_C", gsi_pk="gsi_A", gsi_sk="3"),
    dict(pk="pk_D", gsi_pk="gsi_A", gsi_sk="4"),
    dict(pk="pk_E", gsi_pk="gsi_A", gsi_sk="5"),
    dict(pk="pk_F", gsi_pk="gsi_B", gsi_sk="1"),
    dict(pk="pk_G", gsi_pk="gsi_B", gsi_sk="2"),
]

reversed_pk_data = [
    dict(pk="pk_G", gsi_pk="gsi_A", gsi_sk="1"),
    dict(pk="pk_F", gsi_pk="gsi_A", gsi_sk="2"),
    dict(pk="pk_E", gsi_pk="gsi_A", gsi_sk="3"),
    dict(pk="pk_D", gsi_pk="gsi_A", gsi_sk="4"),
    dict(pk="pk_C", gsi_pk="gsi_A", gsi_sk="5"),
    dict(pk="pk_B", gsi_pk="gsi_B", gsi_sk="1"),
    dict(pk="pk_A", gsi_pk="gsi_B", gsi_sk="2"),
]

random_pk_data = [
    dict(pk=str(uuid4()), gsi_pk="gsi_A", gsi_sk="1"),
    dict(pk=str(uuid4()), gsi_pk="gsi_A", gsi_sk="2"),
    dict(pk=str(uuid4()), gsi_pk="gsi_A", gsi_sk="3"),
    dict(pk=str(uuid4()), gsi_pk="gsi_A", gsi_sk="4"),
    dict(pk=str(uuid4()), gsi_pk="gsi_A", gsi_sk="5"),
    dict(pk=str(uuid4()), gsi_pk="gsi_B", gsi_sk="1"),
    dict(pk=str(uuid4()), gsi_pk="gsi_B", gsi_sk="2"),
]


@pytest.fixture()
def empty_table():

    with mock_aws():
        ddb = boto3.session.Session().resource("dynamodb", region_name="ca-central-1")

        table_name = "test-table"

        table = ddb.create_table(
            TableName=table_name,
            AttributeDefinitions=[
                dict(AttributeName="pk", AttributeType="S"),
                dict(AttributeName="gsi_pk", AttributeType="S"),
                dict(AttributeName="gsi_sk", AttributeType="S"),
            ],
            KeySchema=[
                dict(AttributeName="pk", KeyType="HASH"),
            ],
            GlobalSecondaryIndexes=[
                dict(
                    IndexName="gsi",
                    KeySchema=[
                        dict(AttributeName="gsi_pk", KeyType="HASH"),
                        dict(AttributeName="gsi_sk", KeyType="RANGE"),
                    ],
                    Projection=dict(ProjectionType="ALL"),
                )
            ],
            ProvisionedThroughput=dict(ReadCapacityUnits=100, WriteCapacityUnits=100),
        )

        yield table


@pytest.fixture()
def table_sorted_pk(empty_table):

    with empty_table.batch_writer() as batch:
        for item in sorted_pk_data:
            batch.put_item(Item=item)

    return empty_table


@pytest.fixture()
def table_reversed_pk(empty_table):

    with empty_table.batch_writer() as batch:
        for item in reversed_pk_data:
            batch.put_item(Item=item)

    return empty_table


@pytest.fixture()
def table_random_pk(empty_table):

    with empty_table.batch_writer() as batch:
        for item in random_pk_data:
            batch.put_item(Item=item)

    return empty_table


# Test always passes
def test_exclusive_start_key_w_gsi_sorted_pk(table_sorted_pk):
    """Verify that the exclusive start key is used."""
    query_kwargs = dict(
        IndexName="gsi", KeyConditionExpression=Key("gsi_pk").eq("gsi_A"), Limit=3
    )
    resp = table_sorted_pk.query(**query_kwargs)

    assert [item["gsi_sk"] for item in resp["Items"]] == ["1", "2", "3"]
    assert resp["LastEvaluatedKey"] == sorted_pk_data[2]

    resp = table_sorted_pk.query(
        **query_kwargs | dict(ExclusiveStartKey=resp["LastEvaluatedKey"])
    )

    assert [item["gsi_sk"] for item in resp["Items"]] == ["4", "5"]
    assert "LastEvaluatedKey" not in resp


# Test always fails
def test_exclusive_start_key_w_gsi_reversed_pk(table_reversed_pk):
    """Verify that the exclusive start key is used."""
    query_kwargs = dict(
        IndexName="gsi", KeyConditionExpression=Key("gsi_pk").eq("gsi_A"), Limit=3
    )
    resp = table_reversed_pk.query(**query_kwargs)

    assert [item["gsi_sk"] for item in resp["Items"]] == ["1", "2", "3"]
    assert resp["LastEvaluatedKey"] == reversed_pk_data[2]

    resp = table_reversed_pk.query(
        **query_kwargs | dict(ExclusiveStartKey=resp["LastEvaluatedKey"])
    )

    assert [item["gsi_sk"] for item in resp["Items"]] == ["4", "5"]
    assert "LastEvaluatedKey" not in resp


# Test may or may not fail depending on pk order
def test_exclusive_start_key_w_gsi_random_pk(table_random_pk):
    """Verify that the exclusive start key is used."""
    query_kwargs = dict(
        IndexName="gsi", KeyConditionExpression=Key("gsi_pk").eq("gsi_A"), Limit=3
    )
    resp = table_random_pk.query(**query_kwargs)

    assert [item["gsi_sk"] for item in resp["Items"]] == ["1", "2", "3"]
    assert resp["LastEvaluatedKey"] == random_pk_data[2]

    resp = table_random_pk.query(
        **query_kwargs | dict(ExclusiveStartKey=resp["LastEvaluatedKey"])
    )

    assert [item["gsi_sk"] for item in resp["Items"]] == ["4", "5"]
    assert "LastEvaluatedKey" not in resp
```

</details>

The first test, with the primary key values matching the GSI sort key values, always passes.

The second test, with the primary key values in reverse order wrt the GSI sort key values, always fails.

The third test, with randomized primary key values may or may not pass depending on the order of the primary key values.

Results from one example run below:

<details>
<summary><b>Test run results</b></summary>

```
================================================================ FAILURES ================================================================= 
_______________________________________________ test_exclusive_start_key_w_gsi_reversed_pk ________________________________________________ 

table_reversed_pk = dynamodb.Table(name='test-table')

    def test_exclusive_start_key_w_gsi_reversed_pk(table_reversed_pk):
        """Verify that the exclusive start key is used."""
        query_kwargs = dict(
            IndexName="gsi", KeyConditionExpression=Key("gsi_pk").eq("gsi_A"), Limit=3
        )
        resp = table_reversed_pk.query(**query_kwargs)

        assert [item["gsi_sk"] for item in resp["Items"]] == ["1", "2", "3"]
        assert resp["LastEvaluatedKey"] == reversed_pk_data[2]

        resp = table_reversed_pk.query(
            **query_kwargs | dict(ExclusiveStartKey=resp["LastEvaluatedKey"])
        )

>       assert [item["gsi_sk"] for item in resp["Items"]] == ["4", "5"]
E       AssertionError: assert ['1', '2', '3'] == ['4', '5']
E
E         At index 0 diff: '1' != '4'
E         Left contains one more item: '3'
E
E         Full diff:
E           [
E         -     '4',...
E
E         ...Full output truncated (9 lines hidden), use '-vv' to show

tests\test_moto.py:137: AssertionError
________________________________________________ test_exclusive_start_key_w_gsi_random_pk _________________________________________________ 

table_random_pk = dynamodb.Table(name='test-table')

    def test_exclusive_start_key_w_gsi_random_pk(table_random_pk):
        """Verify that the exclusive start key is used."""
        query_kwargs = dict(
            IndexName="gsi", KeyConditionExpression=Key("gsi_pk").eq("gsi_A"), Limit=3
        )
        resp = table_random_pk.query(**query_kwargs)

        assert [item["gsi_sk"] for item in resp["Items"]] == ["1", "2", "3"]
        assert resp["LastEvaluatedKey"] == random_pk_data[2]

        resp = table_random_pk.query(
            **query_kwargs | dict(ExclusiveStartKey=resp["LastEvaluatedKey"])
        )

>       assert [item["gsi_sk"] for item in resp["Items"]] == ["4", "5"]
E       AssertionError: assert [] == ['4', '5']
E
E         Right contains 2 more items, first extra item: '4'
E
E         Full diff:
E         + []
E         - [
E         -     '4',
E         -     '5',
E         - ]

tests\test_moto.py:156: AssertionError
========================================================= short test summary info ========================================================= 
FAILED tests/test_moto.py::test_exclusive_start_key_w_gsi_reversed_pk - AssertionError: assert ['1', '2', '3'] == ['4', '5']
FAILED tests/test_moto.py::test_exclusive_start_key_w_gsi_random_pk - AssertionError: assert [] == ['4', '5']
```

</details>

Hoping this can help pinpointing the issue.

**Comment by bblommers** [comment]
Hi @ffjdm  and @DrGFreeman , thanks for raising this. This should already have been fixed with https://github.com/getmoto/moto/pull/7835 - can you try and install moto >= 5.0.12.dev11 to verify it works for you?

**Comment by DrGFreeman** [comment]
> Hi @ffjdm and @DrGFreeman , thanks for raising this. This should already have been fixed with #7835 - can you try and install moto >= 5.0.12.dev11 to verify it works for you?

@bblommers Yep, the three [above tests](https://github.com/getmoto/moto/issues/7888#issuecomment-2248788957) pass with moto==5.0.12.dev11.

Thanks! 🚀

**Comment by ffjdm** [comment]
@bblommers Same here, 5.0.12.dev11 fixes the issue, the next page is properly retrieved.
Thank you very much! 💪

**Comment by ffjdm** [comment]
Fixed in 5.0.12.dev11
